Merge from dev to master 17.04.2023

ZhosanDenis opened the pull request.

Garik-create committed:
Частично выполнен функционал по испытательному сроку.

ZhosanDenis committed:
Изменения №3 от 17.04.2023
- добавлена аннотация @Transactional к методу addDailyReport() класса DailyReportService.

ZhosanDenis committed:
Merge pull request #21 from Zhoss/master

Merge pull request #21 from Zhoss/dev

ZhosanDenis committed:
Merge pull request #22 from Zhoss/1-feature-denis

Изменения №3 от 17.04.2023

ZhosanDenis committed:
Изменения №4 от 17.04.2023
- в scripts.sql удалены все changeset'ы, кроме первого - в нем созданы все необходимые таблицы. Сделано, чтобы не было путаницы, и сразу были видны все таблицы и колонки.

Garik-create committed:
Merge branch 'dev' into 1-feature-igor

ZhosanDenis committed:
Изменения №5 от 17.04.2023
- в enum-класс Command добавлен метод по поиску команды (enum) по строке;
- в DailyReportController добавлен эндпоинт по удалению отчета и перехватчик исключений;
- в DailyReportService добавлен метод по удалению отчета по id;
- в классе Dog удалена проверка на пустую строку для setFeatures() - особенностей у собаки может не быть;
- в ModelMapper в метод по конвертации Carer в CarerRecord (mapToCarerRecord()) добавлена проверка на наличие собаки у опекуна - при отсутствии выкидывалась NullPointException;
- в scripts.sql удалено ограничение NOT NULL для колонок dog_diet, dog_health, dog_behavior таблицы dogs - при создании отчета (поэтапно через телеграм бота) выкидывалось исключение;
- TelegramBotUpdatesListener:
1. в метод savePhotoFromCarer() добавлена проверка на наличие в БД отчета за сегодняшнюю дату, чтобы опекун не присылал несколько отчетов. Иначе нельзя их получить через контроллер;
2. в методе handleCommand() структура if-else-if заменена на switch-case для улучшения читабельности кода;
3. в методе handleTextMessage() изменено сохранение частей ежедневного отчета (теперь сохраняется без "2)" и др.);
4. метод sendNotification() не изменен (жду ответа от Дмитрия Бизина), но ctrl+shift+L сдвинул строки.

ZhosanDenis committed:
Merge pull request #24 from Zhoss/1-feature-denis

Изменения №4 от 17.04.2023

Garik-create committed:
Merge branch 'dev' into 1-feature-igor

# Conflicts:
#	src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/listener/TelegramBotUpdatesListener.java

Garik-create committed:
- добавил таймер на проверку (каждый день) окончания испытательного срока для уведомления волонтёра.
- добавил в сервисе и репозитории поиск опекуна по айди собаки.
- добавил в сервисе и репозитории поиск договора по айди опекуна.
- добавил дто-шку по собаке, чтобы в контроллере при изменению поля онПробейшн меньше заполнятьь.
- аналогично сделал для поля Теэйкн.
- добавил эндпойт по изменению поля онПробейшн. Пометил его "Новый", чтобы быстрее найти. Соответственно добавил метод в догСервис.
- аналогично сделал для поля Теэйкн.
- добавил таймер на проверку (каждый час), что клиентом пройден испытательный срок .

Garik-create committed:
Merge pull request #26 from Zhoss/1-feature-igor

1 feature igor

ZhosanDenis committed:
Изменения №1 от 19.04.2023
- класс Agreement: добавлено поле probationEndData, а также геттер и сеттер к нему;
- AgreementRecord: добавлено поле probationEndData, а также геттер и сеттер к нему;
- AgreementService: метод findAllAgreementsByDate() заменен на findAllAgreementsWithProbationByDate() - данный метод ищет все договоры, у которых передаваемая дата ранее даты окончания испытательного срока. Метод findAgreementByCarerId() удален, метод findAgreements() добавлен (поиск всех договоров для нужд Timer);
- enum-класс Command: добавлены 5 команд (продление испытательного срока, уведомление опекуна), добавлен boolean-метод, проверяющий наличие передаваемой string-команды в перечне полей command всех enum-команд;
- DailyReportController: добавлен эндпоинт по получению отчетов за определенную дату;
- DailyReportRepository: добавлен метод findDailyReportsByReportDate() по получению отчетов за определенную дату;
- DailyReportService: добавлен метод findDailyReportsByDate() по получению отчетов за определенную дату;
- DogController: добавлены эндпоинты по изменению статусов собаки isTaken и onProbation;
- удалены классы DogRecord2 и DogRecord3;
- DogService: добавлены методы по изменению статусов собаки isTaken и onProbation;
- ModelMapper: добавлено автоматическое определение даты окончания испытательного срока как дата заключения договор + 30 дней;
- scripts.sql: добавлена колонка probation_end_data в таблицу agreements;
- TelegramBotUpdatesListener: изменен способ скачивания фотографий (было вручную - стало через метод Телеграм бота). В метод handleCommand() добавлена проверка на наличие команды в перечне enum-команд, также добавлена обработка команд по окончанию испытательного срока и уведомления опекуна (при вызове метода decideOnProbation()). В метод handleTextMessage() добавлена обработка сообщения от волонтеров с id опекуна о необходимости сообщить опекуну о некачественных отчетах. Scheduled-методы перенесены в класс Timer.
- Timer: добавлены scheduled-методы по уведомлению опекунов (если опекун не присылал отчет за предыдущий день) и волонтеров (если опекун не присылал отчет в течение двух дней), по напоминанию волонтерам об окончании испытательного срока и принятию дальнейшего решения, по уведомлению волонтеров о необходимости ежедневно проверять отчеты.

ZhosanDenis committed:
Merge pull request #27 from Zhoss/1-feature-denis

Изменения №1 от 19.04.2023

ZhosanDenis committed:
Изменения №1 от 20.04.2023
- создан класс-сущность Client для записи статуса клиента (интересует приют для кошек или для собак);
- создан ClientRepository для добавления сущностей в БД;
- в enum-класс Command добавлены команды по выбору приюта для кошек и приюта для собак, команда по запросу контактных данных службы охраны. Также переименованы команды с DOG в названии (теперь PET);
- добавлена схема проезда для приюта для собак (предыдущая схема проезда используется для приюта для кошек);
- в scripts.sql добавлен changeset с созданием таблицы clients;
- в TelegramBotConfiguration переименованы команды согласно изменениям в enum-классе Command;
- TelegramBotUpdatesListener:
1) заинжектен ClientRepository;
2) изменены обработки сообщений от опекунов и из чата волонтеров (теперь каждые сообщения/команды обрабатываются в разных методах в зависимости откуда отправлены);
3) изменен метод по созданию/обновлению чата волонтеров (теперь чат создается в зависимости от нажатой кнопки в чате волонтеров);
4) изменена обработка сообщений от опекуна в зависимости от того, информацию по какому приюту он ищет;
5) создан метод по обработке команд/сообщений из чата волонтеров;
6) в методе по принятию решения об окончании испытательного срока конструкция if-else заменена на switch-case;
7) обработка команды по уведомлению опекуна о некачественном заполнении отчетов перенесена в метод по обработке сообщений из чата волонтеров;
8) добавлена клавиатура с выбором приюта;
9) в методе с созданием клавиатуры "Как взять питомца из приюта" (этап 2 из ТЗ) теперь создается две разных клавиатуры в зависимости от типа приюта.

Garik-create committed:
- добавил AgreementServiceTest (не удалось протестировать метод deleteAgreement(long id))

ZhosanDenis committed:
Изменения №1 от 21.04.2023
- созданы сущности с припиской Cat и Dog, сущности без приписок стали родительскими классами для созданных сущностей;
- для каждой сущности с припиской создан репозиторий;
- сущности с приписками перенесены в отдельный пакет в зависимости от типа животного;
- в некоторые эндпоинты контроллеров добавлены параметры с типом животного для получения нужных данных;
- в классы-record добавлено поле petType;
- в классах-сервисах возвращаемый тип данных зависит от типа животного, закомментированы (временно) методы для работы с телеграм ботом;
- созданы таблицы в БД для сущностей Cat, переименован таблицы для сущностей Dog;
- закомментирован (временно) код в TelegramBotUpdatesListener и Timer для проверки работы контоллеров-сервисов.

KlochkoVitaliy committed:
Сделаны тесты на Controllers

Garik-create committed:
- добавил CarerServiceTest, не получилось протестировать два void метода editCarer() и deleteCarer()
- добавил hashCode, equals в CarerRecord для сравнения в тестах
- заменил в findAllCarers() List<Carer> carerRecords на List<Carer> carers в CarerService для логичности
- закомментировал в CarerService неиспользуемый метод findCarerByDogId(long dogId)

Garik-create committed:
- сделал недоступными неработающие тесты

ZhosanDenis committed:
Изменения №1 от 23.04.2023
- AgreementController: добавлен эндпоинт по изменению даты окончания испытательного срока (для варианта возврата питомца раньше);
- AgreementService: изменен метод по получению всех договоров с действующим испытательным сроком (в зависимости от типа животного), изменен метод по получению всех договоров с заканчивающимся (осталось 2 дня до конца) испытательным сроком (в зависимости от типа животного);
- application.properties: убран параметр, показывающий действия Hibernate;
- CarerService: изменен метод по добавлению опекуна (из телеграм бота) - в зависимости от типа животного, изменен метод по поиску опекуна по id чата (из телеграм бота) - в зависимости от типа животного, удален неиспользующийся метод по поиску опекуна по id питомца;
- DailyReport: добавлено поле Carer carer + геттер/сеттер, переименованы поля с dog на pet;
- DailyReportRecord: переименованы поля с dog на pet;
- DailyReportService: изменен метод по добавлению ежедневного отчета из телеграм бота (в зависимости от типа животного);
- ModelMapper: в методе updateCarer() повторяющийся код вынесен в отдельный private метод update(), изменен метод mapToDailyRecordRecord() для правильного сохранения ежедневного отчета;
- Pet: добавлены public константы для использования на проверках типа животного по всему приложению;
- TelegramBotConfiguration: удалены команды, связанные с кинологами, чтобы опекуны кошек их не видели;
- TelegramBotUpdatesListener: заинжектены CatCarerRepository и CatAgreementRepository, изменен метод по сохранению фотографии и созданию ежедневного отчета (в зависимости от типа животного), добавлен метод selectVolunteerChatId для выбора чата волонтеров, для решения об окончании испытательного срока (метод decideOnProbation()) предусмотрены обработки команд в зависимости от типа животного, в методе handleTextMessage() изменена логика сохранения частей ежедневного отчета;
- Timer: теперь таймер ищет информацию о договорах в обеих таблицах БД (dog_agreements и cat_agreements) и отправляет напоминание о проверке отчетов в оба чата волонтеров (для кошек и собак).

ZhosanDenis committed:
Изменения №2 от 23.04.2023
- подкорректировал DogControllerTest: добавил необходимые параметры, изменил ссылки запросов. Тесты работают.

KlochkoVitaliy committed:
Дополнены тесты на Controllers

Garik-create committed:
- доделал AgreementServiceTest, CarerServiceTest
- добавил DailyReportServiceTest, DogServiceTest
- подправил DogService

Garik-create committed:
Merge pull request #29 from Zhoss/1-feature-igor

- добавил AgreementServiceTest (не удалось протестировать метод delet…

Garik-create committed:
Merge branch 'dev' into 1-feature-Vitaliy

Garik-create committed:
Merge pull request #30 from Zhoss/1-feature-Vitaliy

Сделаны тесты на Controllers

dmitry-bizin commented on pom.xml:
```diff
@@ -58,6 +58,11 @@
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jsr310</artifactId>
+            <version>2.15.0-rc3</version>
```
версию писать не нужно

dmitry-bizin commented on src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/listener/TelegramBotUpdatesListener.java:
```diff
@@ -153,54 +160,68 @@ public int process(List<Update> updates) {
     }
 
+    private void updateVolunteerChat(long newVolunteerChatId, VolunteerChat volunteerChat) {
+        volunteerChat.setId(1);
```
не нужно никогда явно id выставлять

dmitry-bizin commented on src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/listener/TelegramBotUpdatesListener.java:
```diff
@@ -153,54 +160,68 @@ public int process(List<Update> updates) {
+                        byte[] photos = telegramBot.getFileContent(file);
+                        Files.createDirectories(filePath.getParent());
+                        Files.deleteIfExists(filePath);
+                        try (InputStream is = new ByteArrayInputStream(photos);
```
в одну строку можно в файл записать через Files.write()

dmitry-bizin commented on src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/listener/TelegramBotUpdatesListener.java:
```diff
@@ -217,104 +238,123 @@ private void handleCommand(String command,
-        } else if (command.equals(Command.VOLUNTEER_CONFIRM_COMMAND.getCommand())) {
-            String text = "Спасибо за подтверждение заявки";
-            sendPlainText(chatId, text);
+        Pattern probationPattern = Pattern.compile(
```
Лучше сделать константным полем класса

dmitry-bizin commented on src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/listener/TelegramBotUpdatesListener.java:
```diff
@@ -217,104 +238,123 @@ private void handleCommand(String command,
      */
-
     private void handleTextMessage(String message, long chatId, long volunteerChatId) {
         Pattern clientContactPattern = Pattern.compile(
```
Лучше сделать константным полем класса

dmitry-bizin commented on src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/listener/TelegramBotUpdatesListener.java:
```diff
@@ -217,104 +238,123 @@ private void handleCommand(String command,
     private void handleTextMessage(String message, long chatId, long volunteerChatId) {
         Pattern clientContactPattern = Pattern.compile(
                 "^(([А-я]+\\s){2}[А-я]+)(\\s)(\\+\\d{1,7}\\(\\d{3}\\)\\d{7})$"); //паттерн на ФИО и телефон клиента для записи
+        Pattern carerIdPattern = Pattern.compile(
```
Лучше сделать константным полем класса

ZhosanDenis committed:
Изменения №1 от 27.04.2023
- в контроллерах тип принимаемых параметров с типом животных изменен со String на enum-класс PetType. Также по всем сервисам и TelegramBotUpdatesListener изменен тип данных, показывающих тип животного;
- паттерны сделаны константными полями (TelegramBotUpdatesListener, CarerService);
- скачивание/сохранение в файловую систему сервера фотографии из ежедневного отчета производится через Files.write() - TelegramBotUpdatesListener;
- изменен принцип поиска и сохранения чата волонтеров в телеграм (раньше по id с явным указанием, теперь по типу животного);
- сравнение типов животных в блоках if() осуществляется через ==;
- в таблицы БД clients и volunteer_chats добавлено колонка pet_type.

ZhosanDenis committed:
Merge branch 'dev' into 1-feature-denis

# Conflicts:
#	src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/controller/CarerController.java
#	src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/dto/CarerRecord.java
#	src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/service/CarerService.java
#	src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/service/DailyReportService.java
#	src/main/java/pro/sky/teamwork/animalsheltertelegrambotv2/service/DogService.java
#	src/test/java/pro/sky/teamwork/animalsheltertelegrambotv2/controller/DogControllerTest.java

ZhosanDenis committed:
Изменения №2 от 27.04.2023
- удален класс DogService, появившийся в результате слияния.

ZhosanDenis committed:
Merge pull request #28 from Zhoss/1-feature-denis

Изменения №1 от 20.04.2023

ZhosanDenis committed:
Изменения №1 от 30.04.2023
- добавлены json-файлы для тестирования TelegramBotUpdatesListener;
- DogCarer и CatCarer: добавлены сеттеры для списка ежедневных отчетов (для тестирования);
- удален enum-класс ChatType - вместо него в TelegramBotUpdatesListener используется библиотечный (pengrad) enum-класс Chat.Type;
- Timer: изменен private-метод sendNotification() -> теперь дата последнего отчет должна быть точно равна текущей дате минус 1/2 дня. isBefore() не подходил, т.к. (текущая дата - 2) является датой ранее (текущая дата - 1), т.е. всегда будет выполняться только один блок if (первый).

KlochkoVitaliy committed:
Внесены изменения в ControllerTest 30.04.2023

Garik-create committed:
- добавил equals, hashCCode в AgreementRecord, Pet, PetRecord для тестов
- поправил AgreementServiceTest, CarerServiceTest, DailyReportServiceTest, PetServiceTest
- добавил PetType RET в enum для тестов

Garik-create committed:
- исправили тесты

Garik-create committed:
Merge pull request #31 from Zhoss/1-feature-denis

Изменения №1 от 30.04.2023

Garik-create committed:
Merge pull request #32 from Zhoss/1-1-feature-Vitaliy

Внесены изменения в ControllerTest 30.04.2023

Garik-create committed:
Merge pull request #33 from Zhoss/1-feature-igor

- добавил equals, hashCCode в AgreementRecord, Pet, PetRecord для тестов

ZhosanDenis committed:
Изменения №1 от 30.04.2023
- изменены тесты TimerTest - поправлены текущие даты

ZhosanDenis committed:
Merge pull request #34 from Zhoss/1-feature-denis

Изменения №1 от 01.05.2023

ZhosanDenis committed:
Изменения №2 от 01.05.2023
- небольшие корректировки телеграм бота

ZhosanDenis committed:
Merge pull request #35 from Zhoss/1-feature-denis

Изменения №2 от 01.05.2023
